samples/nrf9160/modem_shell/src/cloud/cloud_shell.c

```diff
@@ -0,0 +1,192 @@
+		break;
+	case NRF_CLOUD_EVT_ERROR:
+		mosh_print("NRF_CLOUD_EVT_ERROR");
+		break;
+	case NRF_CLOUD_EVT_SENSOR_DATA_ACK:
+		mosh_print("NRF_CLOUD_EVT_SENSOR_DATA_ACK");
+		break;
+	case NRF_CLOUD_EVT_FOTA_DONE:
+		mosh_print("NRF_CLOUD_EVT_FOTA_DONE");
+		break;
+	case NRF_CLOUD_EVT_RX_DATA:
+		mosh_print("NRF_CLOUD_EVT_RX_DATA");
+		if (((char *)evt->data.ptr)[0] == '{') {
+			/* Not A-GPS data */
+			break;
+		}
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+		err = nrf_cloud_agps_process((char *)evt->data.ptr, evt->data.len);
+		if (!err) {
+			mosh_print("A-GPS data processed");
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+			/* call us back when prediction is ready */
+			k_work_submit(&notify_pgps_work);
+#endif
+			/* data was valid; no need to pass to other handlers */
+			break;
+		}
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		err = nrf_cloud_pgps_process(evt->data.ptr, evt->data.len);
+		if (err) {
+			mosh_error("Error processing P-GPS packet: %d", err);
+		}
+#else
+		if (err) {
+			mosh_print("Unable to process A-GPS data, error: %d", err);
+		}
+#endif
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
+		mosh_warn("Add the device to nRF Cloud and wait for it to reconnect");
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATED:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATED");
+		break;
+	default:
+		mosh_error("Unknown nRF Cloud event type: %d", evt->type);
+		break;
+	}
+}
+
+static void cloud_connect(void)
+{
+	int err;
+
+	struct nrf_cloud_init_param config = {
+		.event_handler = nrf_cloud_event_handler,
+	};
+
+	err = nrf_cloud_init(&config);
+	if (err == -EACCES) {
+		mosh_print("nRF Cloud module already initialized");
+	} else if (err) {
+		mosh_error("nrf_cloud_init, error: %d", err);
+		return;
+	}
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	else { /* Cloud module successfully initialized */
+		/* Done here, the work structure only gets initialized once */
+		k_work_init(&notify_pgps_work, notify_pgps);
+	}
+#endif
+
+	err = nrf_cloud_connect(NULL);
+	if (err == NRF_CLOUD_CONNECT_RES_ERR_ALREADY_CONNECTED) {
+		mosh_print("nRF Cloud connection already established");
+	} else if (err) {
+		mosh_error("nrf_cloud_connect, error: %d", err);
+	} else {
+		mosh_print("Connection to nRF Cloud established");
+		mosh_print("Endpoint: %s", CONFIG_NRF_CLOUD_HOST_NAME);
+	}
+}
+
+static void cloud_disconnect(void)
+{
+	int err = nrf_cloud_disconnect();
+
+	if (err == -EACCES) {
+		mosh_print("Not connected to nRF Cloud");
+	} else if (err) {
+		mosh_error("nrf_cloud_disconnect, error: %d", err);
+		return;
+	}
+}
+
+int cloud_shell(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret = 0;
+
+	/* command = argv[0] = "cloud" */
+	/* sub-command = argv[1]       */
+	if (argc < 2) {
+		goto show_usage;
+	} else if (strcmp(argv[1], "connect") == 0) {
+		cloud_connect();
+	} else if (strcmp(argv[1], "disconnect") == 0) {
+		cloud_disconnect();
```

Comment:
One more thing. We should have a discussion on how to handle subcommands. Whether this manual stuff is fine or should we use shell macros? I think we should move towards shell macros to get auto-fill feature into use.

Reply:
Shell subcommands would work nicely in this case, so I think it would make sense to use those instead.

Reply:
Done